Heartbeat extension

mildass opened the pull request:
Adds heartbeat extension and unit tests for it.

Fixes #58 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tomato42/tlslite-ng/155)
<!-- Reviewable:end -->

tomato42 commented on unit_tests/test_tlslite_extensions.py:
```diff
@@ -1703,7 +1703,7 @@ def test_extData_none_mode(self):
         ext = HeartbeatExtension().create(0)
 
-        self.assertEqual(ext.extData, '\x00')
+        self.assertEqual(ext.extData, b'\x00')
```
you'll need to [rebase and fixup](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) those commits before force pushing them for the CI to pass

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+    def __init__(self):
+        self.contentType = ContentType.heartbeat
+        self.message_type = 0
+        self.payload_length = 0
```
there's no need for that variable, it always needs to equal `len(payload)`

mildass replied:
I thought that it will be useful when we will want reproduce Heartbleed. I mean creating heartbleed request with different specified payload length and payload.

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+        self.payload_length = 0
+        self.payload = bytearray(0)
+        self.padding = bytearray(0)
+        self.padding_length = 0
```
same here, it always needs to equal `len(padding)`

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+        self.padding = bytearray(0)
+        self.padding_length = 0
+
+    def create(self, message_type, payload_length, payload):
```
`payload_length` is redundant since it needs to equal `len(payload)`, on the other hand, we need a way to specify the amount of padding, and that method is missing it

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+    def create(self, message_type, payload_length, payload):
+        self.message_type = message_type
+        self.payload_length = payload_length
+        # dont check size, because of reproducing Heartbleed
```
for reproducing heartbleed we will be using functionality in tlsfuzzer, no need to pollute reference implementation with such things

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+        self.payload_length = payload_length
+        # dont check size, because of reproducing Heartbleed
+        self.payload = payload
+        self.padding_length = 2**14 - payload_length - 3
```
that will always create messages of maximum size, which makes them useless for MTU discovery

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+        # dont check size, because of reproducing Heartbleed
+        self.payload = payload
+        self.padding_length = 2**14 - payload_length - 3
+        self.padding = bytearray(self.padding_length)
```
the padding bytes need to be random if I'm reading the RFC correctly...

(just for clarity: I don't think the randomness should be inside `create()`, it's a general comment on the padding)

mildass replied:
Sorry for that `bytearray()`.
What should I use for creating random padding?

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+        #p.setLengthCheck(2**14)
+        self.message_type = p.get(1)
+        self.payload_length = p.get(2)
+        self.payload = p.get(self.payload_length)
```
p.get will return a big-endian integer, we want a bytearray, just like it was defined in `__init__`

`getVarBytes()` looks to be more appropriate here to me

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1765,74 @@ def parse(self, p):
+        self.payload_length = p.get(2)
+        self.payload = p.get(self.payload_length)
+        #self.padding_length = 2**14 - self.payload_length - 3
+        #self.padding = p.get(self.padding_length)
```
please don't commit in commented-out code

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -118,6 +118,7 @@ def __init__(self, sock):
         self._defragmenter.addStaticSize(ContentType.change_cipher_spec, 1)
         self._defragmenter.addStaticSize(ContentType.alert, 2)
         self._defragmenter.addDynamicSize(ContentType.handshake, 1, 3)
+        self._defragmenter.addDynamicSize(ContentType.heartbeat, 1, 3) # are these values ok?
```
heartbeats are like application data, so they can't be defragmented

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -578,6 +579,12 @@ def _sendMsg(self, msg, randomizeFirstBlock = True):
+                yield result
+            if len(msg.write()) == 0:
+                return
+
```
won't that cause the application to send two hearbeats?

mildass commented:
Why did comments disappear? I´ve done usual process. Rebase, commit and then force push and it looks like that force push was problem, but why? Can I restore your comments?

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -738,6 +756,8 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
             elif recordHeader.type == ContentType.application_data:
                 yield ApplicationData().parse(p)
+            elif recordHeader.type == ContentType.heartbeat:
+                yield Heartbeat().parse(p)
```
if _getMsg handles the messages itself (line 732), why return them to calling application?

to be precise: I have no problem with either option, but not both at the same time...

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +932,49 @@ def _changeWriteState(self):
+                yield result
+        except GeneratorExit:
+            raise
+        except Exception:
```
you shouldn't catch a bare Exception

tomato42 commented on unit_tests/test_tlslite_extensions.py:
```diff
@@ -1679,5 +1680,54 @@ def test_parse_with_trailing_data(self):
+
 if __name__ == '__main__':
-    unittest.main()
+    unittest.main()
```
that's not a needed change

tomato42 commented:
> Why did comments disappear? I´ve done usual process. Rebase, commit and then force push and it looks like that force push was problem, but why? Can I restore your comments?

I made comments on specific commit IDs, so they are assigned to them
if you're rebasing, github tries to preserve them, but it uses the commit message for that - if the commit message is different, it will consider the commits to be different and then not show the messages

tomato42 commented:
also, please fix codeclimate issues, you can ignore the ones about cyclomatic complexity in tlsrecordlayer.py or tlsconnection.py though

mildass commented:
Could you please look at `unit_tests/test_tlslite_messages.py` to the latest test `TestHeartbeat.test_write()`? Im using `assertRegex`, but it makes problems at Python 2.7 CI tests. What should I use instead of it? When I ran `python2.7 -m unittest discover -v`, its ok, so I dont know why it fails here.

tomato42 commented on unit_tests/test_tlslite_messages.py:
```diff
@@ -2454,5 +2455,67 @@ def test_write(self):
+        heartbeat = Heartbeat().create(HeartbeatExtensionMessages.request,
+                                       1, b'\xAA')
+
+        self.assertRegex(heartbeat.write(), b'^\x01' +       # request
```
why not compare `heartbeat.write()[:4]` with the byte literal?

mildass replied:
Yea, its better than that regex :-D Sorry for my weird solution.

mildass commented:
Whats the best way how to solve that codeclimate `_noneAsUnknown` duplication at `tlslite/messages.py`? Should I call this static method from `Alert` or create super class with only this method?

tomato42 commented:
`_noneAsUnknown` probably should be an external utility function - it's useful also for extensions, and having a common superclass between extensions and messages doesn't look nice to me

mildass commented:
So should I move that function to `tlslite/utils`? If yes, then to which file?

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1759,61 @@ def parse(self, p):
+
+    @property
+    def messageType(self):
+        return noneAsUnknown(HeartbeatExtensionMessages.toRepr(
```
nit: doc missing

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1759,61 @@ def parse(self, p):
+                             self.message_type), self.message_type)
+
+    def __str__(self):
+        return "Heartbeat {0}".format(self.messageType)
```
nit: doc missing

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +928,45 @@ def _changeWriteState(self):
         self._recordLayer.changeReadState()
+
+
+    def writeHeartbeat(self, type, payload, padding_length):
```
`type` is a built-in, the code is masking it like this

tomato42 replied:
given the differences between `write`, `writeAsync`, `read` and `readAsync`, it seems to like the more consistent name for this method would be `writeHeartbeatAsync` and for the method below to be called `writeHeartbeat`

tomato42 replied:
this method does send the hearbeat requests - how do the user application be able to know that we received response to the request we sent?

tomato42 replied:
I meant renaming it, but good point that this will always send a request so the type is redundant.

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +928,45 @@ def _changeWriteState(self):
+            for result in self._sendMsg(heartbeat_request, \
+                                        randomizeFirstBlock=True):
+                yield result
+        except GeneratorExit:
```
what's the purpose of this except?

mildass replied:
in this part I have inspired at `writeAsync` and I thought, that heartbeat messages sending should be handled similary. Is it wrong? If yes, why we want it at `writeAsync` and not here? And should I catch here any other exception?

tomato42 replied:
with `writeAsync`, we need to call `_shutdown()` on errors but GeneratorExit is one of the exceptions that is normally raised by generators (functions that use `yield`) and it doesn't mean that the connection needs to be closed

that code probably should use `StandardError`, but since the whole tlsrecordlayer and tlsconnection need to be rewritten, I'm leaving them as is and allowing them to collect dept

mildass replied:
Catching here `StandardError` exception here and raising it is ok? I dont know how we have to handle these problems, but I suppose that we dont have to shutdown connection after error in sending heartbeat, right?

tomato42 replied:
I was talking about `writeAsync()`, I don't see a reason for `try: ... except` construct in this method at all

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +928,45 @@ def _changeWriteState(self):
+            heartbeat_request = Heartbeat().create(type, payload,
+                                                   padding_length)
+
+            for result in self._sendMsg(heartbeat_request, \
```
`\` is redundant inside parentheses

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +928,45 @@ def _changeWriteState(self):
+                                                   padding_length)
+
+            for result in self._sendMsg(heartbeat_request, \
+                                        randomizeFirstBlock=True):
```
`randomizeFirstBlock` will fragment the message, Heartbeat messages MUST NOT be fragmented

in general, since the fragmentation happens inside `_sendMsg`, I'm not convinced that using this method over `_sendMsgThroughSocket` is correct

tomato42 commented on unit_tests/test_tlslite_extensions.py:
```diff
@@ -1679,5 +1680,53 @@ def test_parse_with_trailing_data(self):
+
+        with self.assertRaises(SyntaxError):
+            ext.parse(parser)
+
```
`clientResponse` method coverage missing

mildass replied:
I think that you meant test for `createResponse`, right? It is at `unit_tests/test_tlslite_messages.py`.

tomato42 replied:
yes, `createResponse`

tomato42 commented on tlslite/utils/format_input.py:
```diff
@@ -0,0 +1,21 @@
+#
+# See the LICENSE file for legal information regarding use of this file.
+
+"""Helper functions for format input"""
```
again, those look more like output formatting to me...

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +722,19 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
                                 yield result
                             continue
 
+                    # If we received a hearbeat request ...
```
typo `hearbeat`

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -833,6 +846,9 @@ def _getNextRecord(self):
                 yield (header, parser)
+            # heartbeat message isn't made out of messages, too
+            elif header.type == ContentType.heartbeat:
+                yield (header, parser)
```
when will that code be reached?

mildass replied:
When we are reading messages from socket and we got heartbeat message, then we dont want let it to Defragmenter and just return header and parser. 
You wrote me in email, that I should handle it in `_getNextRecord()` like Application Data, so I inspired by previous condition. Without this when we receive heartbeat message, then it goes throught to Defragmenter and it cant handle it.

tomato42 replied:
d'oh, I was thinking this is in `_getMsg`, yes, for `_getNextRecord` it's exactly what we need

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +722,19 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
                             continue
 
+                    # If we received a hearbeat request ...
+                    if recordHeader.type == ContentType.heartbeat:
```
how do we know that the heartbeat protocol is supported by the server and negotiated for the connection?

mildass replied:
where I can get here, if it is supported? I looked to some variable, but didnt find any, that got informations about negotiated parameters for connection

tomato42 replied:
I mean, if you receive that message type, and you have not negotiated heartbeat through extensions, the response to that should be `unexpected_message`, so we should set some flag in TLSRecordLayer to remind us

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +722,19 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                    if recordHeader.type == ContentType.heartbeat:
+                        try:
+                            heartbeat_request = Heartbeat().parse(p)
+                            #We answer with heartbeat response based on request
```
what if the message is a response to our request?

mildass replied:
Can I get messages, what I have sent? Because if I will find, that we got response, then I dont know, how to check, if it got relevant payload.

tomato42 replied:
the type of the received message will be different, and it needs to be handled differently:
 1. for requests - send reply, continue waiting for next messages
 2. for replies - call the callback, provide the parsed message to it, maybe the socket too (so that one callback can be reused between different sockets)

tomato42 commented on tlslite/constants.py:
```diff
@@ -252,6 +254,20 @@ class CertificateStatusType(TLSEnum):
+    """Types of heartbeat modes"""
+
+    peer_allowed_to_send_mode = 1
+    peer_not_allowed_to_send_mode = 2
```
generally, the names used for the constants are consistent with the names defined in the RFC, why it's not the case here?

tomato42 commented on tlslite/constants.py:
```diff
@@ -252,6 +254,20 @@ class CertificateStatusType(TLSEnum):
     ocsp = 1
 
 
+class HeartbeatExtensionModes(TLSEnum):
```
why not `HeartbeatMode`, as it is in RFC?

tomato42 commented on tlslite/constants.py:
```diff
@@ -252,6 +254,20 @@ class CertificateStatusType(TLSEnum):
+    peer_not_allowed_to_send_mode = 2
+
+
+class HeartbeatExtensionMessages(TLSEnum):
```
why not `HeartbeatMessageType`, as it is in RFC?

tomato42 commented on tlslite/constants.py:
```diff
@@ -252,6 +254,20 @@ class CertificateStatusType(TLSEnum):
+
+
+class HeartbeatExtensionMessages(TLSEnum):
+    """Types of heartbeat messages"""
```
please add the RFC number those definitions come from

tomato42 commented on tlslite/constants.py:
```diff
@@ -107,7 +107,8 @@ class ContentType(TLSEnum):
     handshake = 22
     application_data = 23
-    all = (20, 21, 22, 23)
+    heartbeat = 24
```
please add the RFC number it comes from as a comment (see line 137 below as an example)

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +928,45 @@ def _changeWriteState(self):
+        @raise socket.error: If a socket error occurs.
+        """
+        try:
+            if self.closed:
```
what if the peer didn't advertise that we're allowed to send heartbeat requests?

tomato42 commented:
I was also wondering about how to handle the responses you receive to the requests you sent - I think using a callback mechanism on the tlslite-ng level would be cleanest - with the callback being called every time the response is received

mildass commented on tlslite/utils/format_input.py:
```diff
@@ -0,0 +1,21 @@
+# Authors:
+#   Hubert Kario
```
Wasn´t sure, what name I have to write here, because you are author of that function

tomato42 replied:
but you've refactored it, and you've created the file so your authorship should be there

mildass commented on tlslite/messages.py:
```diff
@@ -19,6 +19,7 @@
 from .x509certchain import X509CertChain
 from .utils.tackwrapper import *
 from .extensions import *
+from .utils.format_input import noneAsUnknown
```
So `format_output` is appropriate name? Ive got problem with creating suitable names for files/variable...

tomato42 replied:
yes, `format_output` does look more appropriate to me

>Ive got problem with creating suitable names for files/variable...

everybody does :)

mildass commented:
Can you please describe your idea about using callback mechanism? Ive read how it works, but I have never done it in code before and I have no idea, how it should look like here. If I just call function, after receiving response and check, if its ok to our request I have send, but for this we dont need callback or am I wrong?

mildass commented on tlslite/messages.py:
```diff
@@ -1765,3 +1759,61 @@ def parse(self, p):
+        """Create heartbeat request or response with selected parameters"""
+        self.message_type = message_type
+        self.payload = payload
+        if not(15 < padding_length < 2**14 - len(self.payload) - 3):
```
QuantifiedCode recommended me this format for condition `if not( _ < variable < _):`, so what format would be ok? 
And should I just delete this condition and dont check it here, right?

tomato42 replied:
for one, `not` is not a function, it's a reserved keyword, like `if`, `else`, `or`, etc. so there needs to be a space between it and `(`

I have checked if `not (_ < variable <_)` behaves the same as `not _ < variable < _` and they seemed to be exactly equivalent, but maybe I missed some case...

but true, all of it is moot given that this test shouldn't be here in the first place

mildass commented on tlslite/messages.py:
```diff
@@ -1765,3 +1759,61 @@ def parse(self, p):
+    def createResponse(self):
+        """Creates heartbeat response based on request."""
+        heartbeat_response = Heartbeat().create(
+                                        HeartbeatExtensionMessages.response,
```
How it should look like?
`
heartbeat_response = Heartbeat().create(
    HeartbeatExtensionMessages.response,
    self.payload,
    16)
`
is ok?

mildass commented on tlslite/handshakesettings.py:
```diff
@@ -161,6 +161,7 @@ def __init__(self):
         self.requireExtendedMasterSecret = False
         self.dhParams = None
         self.dhGroups = list(ALL_DH_GROUP_NAMES)
+        self.useHeartbeatExtension = True
```
What `verify()` you mean?

tomato42 replied:
ah, sorry, I meant `validate()`, in the same class

tomato42 commented:
> Can you please describe your idea about using callback mechanism? Ive read how it works, but I have never done it in code before and I have no idea, how it should look like here.

```python
from __future__ import print_function

def some_func_cb(a, b):
    print("In the callback: {0}, {1}".format(a, b))
    b += bytearray(b'qux')

class BigObject(object):
    def __init__(self):
        self.cb = None

    def do_stuff(self):
        foo = 12
        bar = bytearray("abba")
        if self.cb:
            self.cb(foo, bar)
        print("in the do_stuff: {0}, {1}".format(foo, bar))

obj = BigObject()
obj.do_stuff()

obj = BigObject()
obj.cb = some_func_cb
obj.do_stuff()
```
executing that will print:
```
in the do_stuff: 12, abba
In the callback: 12, abba
in the do_stuff: 12, abbaqux
```

> If I just call function, after receiving response and check, if its ok to our request I have send, but for this we dont need callback or am I wrong?

1. the server does not have to respond to a hearbeat request
2. you can have multiple hearbeat requests and responses travelling at the same time
3. the server can respond to requests asynchronously - first provide you with a megabyte of data and only then the heartbeat response

at the same time, for applications it is useful to know when the response to heartbeat arrived, but we can't inform the application using the regular means (return values of `read()` or `write()`) as that would break API, and requiring the application to poll tlslite-ng would make it inefficient, callbacks allow us to use the event based approach and delegate the garbage collection of the information about sent requests to the application

BTW: Given that we'll need to add some more logic to handling of heartbeat requests and responses, I think it would be better to separate the changes that add parsing and serialising of heartbear messages and extensions in one commit while the logic that governs their use to another.

mildass commented:
About that last note - how you want separate these changes? Should I create new branch, add to it just parsing and serializing and new pull request with it, then you will merge it and I will create next pull request with handling sending/receiving heartbeat messages? Or something else? Because Im not sure, what else can I do, when I rebased it and last commits were `git commit --amend`

tomato42 commented:
Using `git rebase --interactive`: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +722,19 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                    if recordHeader.type == ContentType.heartbeat:
+                        try:
+                            heartbeat_request = Heartbeat().parse(p)
+                            #We answer with heartbeat response based on request
```
comments should start with `# `

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +722,19 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
                             continue
 
+                    # If we received a heartbeat request ...
+                    if recordHeader.type == ContentType.heartbeat:
```
if that type was received but the extension was not negotiated, it should cause connection abort through `unexpected_message`

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +722,19 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                        try:
+                            heartbeat_request = Heartbeat().parse(p)
+                            #We answer with heartbeat response based on request
+                            heartbeat_response = heartbeat_request.\
```
responses should be sent only if the received heartbeat message is a request

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +925,40 @@ def _changeWriteState(self):
+        """Start a write operation of heartbeat_request.
+
+        @type payload_length: int
+        @param payload_length: Length of payload.
```
no `payload_length` parameter in method

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +925,40 @@ def _changeWriteState(self):
+        """Sends heartbeat_request.
+
+        @type payload_length: int
+        @param payload_length: Length of payload.
```
no `pyload_length` parameter in the method

tomato42 commented on tlslite/constants.py:
```diff
@@ -135,6 +136,7 @@ class ExtensionType(TLSEnum):
     supports_npn = 13172
     tack = 0xF300
     renegotiation_info = 0xff01  # RFC 5746
+    heartbeat = 15  # RFC 6520
```
nit: ids are sorted by their numerical value

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1759,64 @@ def parse(self, p):
+        """Deserialize heartbeat message from parser."""
+        self.message_type = p.get(1)
+        self.payload = p.getVarBytes(2)
+        return self
```
saying in a comment that we're explicitly ignoring leftover bytes (padding) would make it a bit easier to know what's happening

tomato42 commented on unit_tests/test_tlslite_extensions.py:
```diff
@@ -1679,5 +1680,53 @@ def test_parse_with_trailing_data(self):
+        self.assertIsNone(ext.mode)
+
+    def test_create(self):
+        ext = HeartbeatExtension().create(0)
```
this should use the constant so that we're not only testing that the values are passed through, but that the constants are correct

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -1334,6 +1340,11 @@ def _handshakeServerAsyncHelper(self, verifierDB,
             extensions.append(ECPointFormatsExtension().create(
                 [ECPointFormat.uncompressed]))
 
+        if clientHello.getExtension(ExtensionType.heartbeat):
```
what if use of heartbeats is not enabled in settings?

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -773,6 +776,9 @@ def _clientGetServerHello(self, settings, clientHello):
                         "Server selected ALPN protocol we did not advertise"):
                     yield result
+        if settings.useHeartbeatExtension and \
+                serverHello.getExtension(ExtensionType.heartbeat):
```
what if server didn't allow us to send requests? (`peer_not_allowed_to_send`)
what if the value server sent is invalid?

mildass replied:
How can I get, if the mode, what I got from opposite site, is in `HeartbeatMode` class? I mean is any best practice, how I can iterate through class atributes and compare them with value?

tomato42 replied:
> How can I get, if the mode, what I got from opposite site, is in HeartbeatMode class?

it if isn't `toRepr()` will return `None`, but since there can be just two values here, each having a different code-path associated with it, having a default `else` case should be both enough and cleaner

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -147,6 +147,9 @@ def __init__(self, sock):
         self.recordSize = 16384 # 2**14
 
+        # Indicator for heartbeat extension
+        self.heartbeat = False
```
what if we can't send requests but server can? I think one variable is not enough to keep the state

mildass replied:
Variables with names `heartbeat` for our mode and `heartbeatOppositeSide` for other side is ok? Is this better, than `heartbeatClientMode` and `heartbeatServerMode`, in which we dont know, which mode is our, until we find if we are client/server?

tomato42 replied:
I was thinking of `heartbeatCanReceive` and `heartbeatCanSend`

remember that those values are used by both server and client, but the meaning is revered for `heartbeatClientMode` if you're server

tomato42 replied:
oh, and probably `heartbeatSupported` is also needed (as, IIRC, receiving heartbeat if the extension was negotiated but peers are not allowed to send requests should not cause the connection to be dropped)

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +725,31 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                                yield result
+
+                        try:
+                            heartbeat_message = Heartbeat().parse(p)
```
what's the expected behaviour if parsing fails?

mildass replied:
At `_getMsg` is `try:`, that is catching exceptions raised by a Parser or Message instances and sending `decode_error` alert. Or we want something different, when we fail at parsing heartbeat message?

tomato42 replied:
If I remember correctly, the RFC says that malformed heartbeat messages are to be ignored, isn't that the case?

mildass replied:
I found 3 parts talking about when heartbeat messages are to be ignored:
- payload_length of a received HeartbeatMessage is too large
- received HeartbeatResponse message does not contain the expected payload
- it arrives during the handshake
Im not sure, but I think that none of points is this case. I was thinking when parser can fail and he can fail when theres nothing to parse, no payload or different payload and payload length information.

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +725,31 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                        # If heartbeat extension was not negotiated, then
+                        # we send unexpected_message fatal alert
+                        if not self.heartbeat:
+                            for result in self._sendError( \
```
nit: backslash is unnecessary between brackets

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +725,31 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                        if not self.heartbeat:
+                            for result in self._sendError( \
+                                    AlertDescription.unexpected_message,
+                                            "Heartbeat was not negotiated"):
```
nit: unaligned indents

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +725,31 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                                    createResponse()
+                                for result in self._sendMsg(
+                                        heartbeat_response):
+                                    yield result
```
and what happens if the heartbeat message is a response?
what if it is neither?

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +725,31 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                        # we send unexpected_message fatal alert
+                        if not self.heartbeat:
+                            for result in self._sendError( \
+                                    AlertDescription.unexpected_message,
```
if heartbeats are not enabled, won't this be handled by the default case below (the one with `#Otherwise`) if we enter the ContentType.hearbeat case only if heartbeats were negotiatied?

tomato42 commented:
also, in tlsconnection.py, handling of the extension during session resumption (see line 1629) is missing

sanity checking of extension in _serverGetClientHello() is missing

mildass commented:
Is good idea to add variable with heartbeat mode to `HandshakeSettings`? Value, that will be used every time(dont matter if we are client/server) when we will use heartbeat extension.

tomato42 commented:
> Is good idea to add variable with heartbeat mode to HandshakeSettings? Value, that will be used every time(dont matter if we are client/server) when we will use heartbeat extension.

we'll always send responses if the extension was negotiated, so I'm not sure what purpose would that serve...

mildass commented:
About session resumption - for this I have to add heartbeat to `Session`, right? At `Session` is `create()` that is being used for creating session.
My question is: how should I handle this session creating in tlslite-ng? Add default value for creating session as `heartbeat=True` or as `False` and then change sessios creating in other places(e.g. at `TLSConnection` in server part is being created session, should I add heartbeat here)?

tomato42 commented:
> About session resumption - for this I have to add heartbeat to Session, right?

does the RFC state that you have to negotiate it for the whole session, not just for the connection?

mildass commented:
Im not sure, what do you mean. At Heartbeat RFC I found only: "Usually, such protocols have no session management."
Is it means, that I dont have to save it to session? Because I thought that I have to add it, when you wrote
> also, in tlsconnection.py, handling of the extension during session resumption (see line 1629) is missing

tomato42 commented:
> Im not sure, what do you mean. At Heartbeat RFC I found only: "Usually, such protocols have no session management."

[Section 2 of RFC 6520](https://tools.ietf.org/html/rfc6520#section-2):
```
   The support of Heartbeats is indicated with Hello Extensions. [...]
   This decision can be changed with every renegotiation.
```

> Does it mean, that I dont have to save it to session? Because I thought that I have to add it, when you wrote


Session resumption behaviour seems to be undefined, but given that it can be changed with every renegotiation, I'd say that it needs to be negotiated anew with every Client Hello/Server Hello exchange - that includes resumption

what I meant, is that server will not send its heartbeat extension in resumed session, and it should

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +735,40 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                            elif heartbeat_message.message_type == \
+                                    HeartbeatMessageType.heartbeat_response:
+                                heartbeat_message.checkResponse(
+                                    self.checkHeartbeatResponse)
```
tlsrecordlayer.py is part of the library, what I thought about callbacks, is that they should be provided by the application that uses the library, in our case that would be either the integration test cases in `tests/tlstest.py` or example server/client in `scripts/tls.py`

mildass replied:
Ah, ok. But we still need some variable, which will be used as argument here, and application before starting server/client can set this variable to function, right? And user must know, that variable can be set to function (not to value) and how function header must look like, yes?

tomato42 replied:
> But we still need some variable, which will be used as argument here

yes, the callback itself should be enough (as in, if we don't have callback defined, we do not allow other side to send heartbeat requests)

> how function header must look like

yes, but that header needs only to be in documentation string - there's no strict type checking in python

mildass replied:
Variable for callback in `HandshakeSettings` is ok? I don't know in which other classes are being variables set before start, but its not variable for handshake, so not sure if its good way.

tomato42 replied:
yes, it should be in HandshakeSettings. 

> its not variable for handshake

well, it does control what value we send in the heartbeat extension :)

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +962,53 @@ def _changeWriteState(self):
+        if self.closed:
+            raise TLSClosedConnectionError(
+                "attempt to write to closed connection")
+        heartbeat_request = Heartbeat().create(
```
shouldn't we also check here if the heartbeat is actually negotiated?

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -648,6 +648,10 @@ def _clientSendClientHello(self, settings, session, srpUsername,
             extensions.append(ALPNExtension().create(alpn))
+        if settings.useHeartbeatExtension:
+            extensions.append(HeartbeatExtension().create(
+                HeartbeatMode.peer_allowed_to_send))
```
that setting should depend on the callback being registered or not

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -773,6 +777,24 @@ def _clientGetServerHello(self, settings, clientHello):
+                        "Server sent Heartbeat extension without one in " +
+                        "client hello"):
+                    yield result
+            heartbeatExt = serverHello.getExtension(ExtensionType.heartbeat)
```
wouldn't it be cleaner to call `getExtension()` just once?

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -1660,6 +1702,25 @@ def _serverGetClientHello(self, settings, certChain, verifierDB,
+                                "Client sent invalid Heartbeat extension"):
+                            yield result
+                    heartbeat = HeartbeatExtension().create(
+                        HeartbeatMode.peer_allowed_to_send)
```
again, this probably should depend on the callback being registered

mildass replied:
What do you mean with it? Should I create some attribute in `HeartbeatExtension` class and `Heartbeat` class will work with it?

tomato42 replied:
No, sorry, too many combinations. When we're a server, and are replying to a request, we don't need any callbacks, we can always respond to hearbeats.

tomato42 commented on tlslite/messages.py:
```diff
@@ -1765,3 +1759,82 @@ def parse(self, p):
+            16)
+        return heartbeat_response
+
+    def checkResponse(self, callback=None):
```
code in `messages.py` and `extensions.py` is just parsing and decoding values, it does not perform any kind of validation or verification outside of checking if the message is well formed.

For now I'm putting that code in [tlsconnection.py](https://github.com/tomato42/tlslite-ng/blob/master/tlslite/tlsconnection.py#L1486), later on it will need to be moved in a separate module, but definitely not here.

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -773,6 +777,29 @@ def _clientGetServerHello(self, settings, clientHello):
+            if not settings.useHeartbeatExtension:
+                for result in self._sendError(
+                        AlertDescription.unsupported_extension,
+                        "Server sent Heartbeat extension without one in " +
```
the `+` is unnecessary

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -1542,6 +1577,23 @@ def _serverGetClientHello(self, settings, certChain, verifierDB,
+            else:
+                for result in self._sendError(
+                        AlertDescription.illegal_parameter,
+                        "Client sent invalid Heartbeat extension"):
```
How about "Received invalid value in Hearbeat extension"?

tomato42 replied:
the problem tough is when the client sent "peer_allowed_to_send" but the callback is not set, then the exception is raised...

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -147,6 +147,20 @@ def __init__(self, sock):
+        # Indicator, that both sides want use heartbeat extension
+        self.heartbeatSupported = False
+
+        # Callback function for receiving Heartbeat requests
```
isn't it for "handling responses to hearbeat requests we sent"?

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -722,6 +736,40 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                            # request we sent
+                            elif heartbeat_message.message_type == \
+                                    HeartbeatMessageType.heartbeat_response:
+                                heartbeat_message.checkResponse(
```
the callback could be called here directly

mildass replied:
What do you mean? Directly calling e.g. `self.heartbeatReponceCallback(heartbeat_message.payload)`? I´m not sure about it and how callback can look like (what parameters it will have).

tomato42 replied:
either just payload, or the whole `hearbeat_message`

the parameter of that callback would then be just that - the response received.

it can look like in any possible way after that, quick example:
```
sent_heartbeats = []
heartbeat_count = 0


def heartbeat_cb(message):
   no, payload = next((no, payload) for no, payload in sent_requests if payload == message.payload), (None, None))
   if no is None:
     raise ValueError("Received unexpected heartbeat response")
  sent_requests.remove((no, payload))
  print("Got response to hearbeat # {0}".format(no))
  
def main():
   ...
   handshake_settings = HandshakeSettings()
   handshake_settings.heartbeat_callback = heartbeat_cb
   socket.handshake(handshake_settings)

   payload = getRandomBytes(32)
   heartbeat_count += 1
   sent_heartbeats.append((heartbeat_count, payload))
   socket.sentHeartbeatRequest(payload)
   ...
```

(sorry for typos, mismatched object names and method names, writing from memory, hope it's useful)

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -912,3 +963,45 @@ def _changeWriteState(self):
+
+        @raise socket.error: If a socket error occurs.
+        """
+        self.heartbeatRequestPayloads.append(payload)
```
leftover of the non-callback code?

tomato42 commented on unit_tests/test_tlslite_messages.py:
```diff
@@ -2454,5 +2455,72 @@ def test_write(self):
+
+        self.assertEqual("unknown(5)", heartbeat.messageType)
+
+    def test___str___(self):
```
extra `_` at the end :)

the name format is `test_` + name of function. Since function name is `__str__` the name of test should be `test___str__`

mildass commented:
Updated to master and added test case for heartbeat and callback, but I´m not sure, if test cases flow is ok.

And I´m getting timeout on that new test case and I don´t know why.  I have tried just adding new test case with simple flow (synchro, connect, handshake, close connection), to see if it is ok, but it´s still giving me timeout too. Do you have any ideas, what is wrong?

tomato42 commented on tests/tlstest.py:
```diff
@@ -677,6 +677,20 @@ def connect():
+    settings = HandshakeSettings()
+    heartbeatPayload = os.urandom(100)
+    def heartbeatResponseCheck(message):
+        assert(heartbeatPayload == message.payload)
```
nit: assert is a statement, doesn't require `()`

tomato42 commented on tests/tlstest.py:
```diff
@@ -677,6 +677,20 @@ def connect():
 
     test_no += 1
 
+    print("Test {0} - Heartbeat extension response callback".format(test_no))
```
server side configuration missing

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -1628,6 +1663,21 @@ def _serverGetClientHello(self, settings, certChain, verifierDB,
+        heartbeatExt = clientHello.getExtension(ExtensionType.heartbeat)
+        if heartbeatExt:
+            if heartbeatExt.mode == HeartbeatMode.peer_allowed_to_send and \
+                    settings.heartbeatResponseCallback:
```
case with "we know how to handle heartbeats, but we will never send them" is missing
(in other words, check for `peer_allowed_to_send` should happen in isolation)

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -729,6 +744,40 @@ def _getMsg(self, expectedType, secondaryType=None, constructorType=None):
+                                    for result in self._sendError(
+                                            AlertDescription.
+                                            unexpected_message,
+                                            "Received heartbeat_request to " +
```
nit: `+` redundant inside `()`

tomato42 commented:
> I´m getting timeout on that new test case and I don´t know why.

I'm getting same thing Travis is getting:
```
Test 0 - Anonymous server handshake
Traceback (most recent call last):
  File "./tlstest.py", line 1399, in <module>
    serverTestCmd(sys.argv[2:])
  File "./tlstest.py", line 823, in serverTestCmd
    connection.handshakeServer(anon=True)
  File "/home/hkario/dev/tlslite-1/tlslite/tlsconnection.py", line 1219, in handshakeServer
    nextProtos=nextProtos, anon=anon, alpn=alpn, sni=sni):
  File "/home/hkario/dev/tlslite-1/tlslite/tlsconnection.py", line 1248, in handshakeServerAsync
    for result in self._handshakeWrapperAsync(handshaker, checker):
  File "/home/hkario/dev/tlslite-1/tlslite/tlsconnection.py", line 2287, in _handshakeWrapperAsync
    for result in handshaker:
  File "/home/hkario/dev/tlslite-1/tlslite/tlsconnection.py", line 1291, in _handshakeServerAsyncHelper
    anon, alpn, sni):
  File "/home/hkario/dev/tlslite-1/tlslite/tlsconnection.py", line 1679, in _serverGetClientHello
    "Received invalid value in Heartbeat extension"):
  File "/home/hkario/dev/tlslite-1/tlslite/tlsrecordlayer.py", line 575, in _sendError
    raise TLSLocalAlert(alert, errorStr)
tlslite.errors.TLSLocalAlert: illegal_parameter: Received invalid value in Heartbeat extension
```
so I'm not sure how you're able to run the case you have added... It's because the server side is missing a case in an `if`, see above comments.

> And I´d like to ask about default behavior - if I remember well, we talked that Heartbeat should be supported in default settings,

yes, a peer should be able to _answer_ heartbeats by default

> but then we need to have set default callback function.
> So what behavior should have that Heartbeat callback function and in which file it should be? (that´s why tests are failing now, because I set that Heartbeat should be used, but didnt defined callback func.)

callback is needed only when we _send_ heartbeats, so there is no need for default callback function

mildass commented:
Why travis-cl shows "Waiting for status to be reported"? What can I do with it?

tomato42 commented:
well, the PR has conflicting files, so it's unmergeable, you have to rebase it, then CI will re-run automatically

tomato42 commented:
@mildass please, [rebase](https://help.github.com/articles/about-git-rebase/), do **not** "branch update"

mildass commented:
Yeah, sorry. I tried "resolve conficts" option at github.com, but it made new commit about merging.

tomato42 commented on tests/tlstest.py:
```diff
@@ -735,6 +735,20 @@ def connect():
+    settings.heartbeatResponseCallback = heartbeatResponseCheck
+    connection.handshakeClientCert(serverName=address[0], settings=settings)
+    connection.sendHeartbeatRequest(heartbeatPayload, 16)
+    connection.close()
```
`testConnClient(connection)` missing

tomato42 replied:
it's still misssing...

tomato42 commented on tests/tlstest.py:
```diff
@@ -1405,6 +1419,21 @@ def server_bind(self):
+    connection.handshakeServer(certChain=x509Chain, privateKey=x509Key,
+                               settings=settings)
+    connection.sendHeartbeatRequest(heartbeatPayload, 16)
+    connection.close()
```
`testConnServer(connection)` missing

tomato42 replied:
still missing

tomato42 commented:
something's broken (see travis-ci):
```
    raise TLSInternalError("attempt to send Heartbeat request when "
tlslite.errors.TLSInternalError: TLSInternalError('attempt to send Heartbeat request when we cant send it to other side',
```

mildass commented:
It's weird, because it was ok and now I'm getting same error as travis-ci or timeout, when I add mentioned `testConnServer(connection)` before sending heartbeat request.

Wireshark shows this (after ClientHello, ServerHello etc.): 
clients sends ClientKeyExchange, CCS and encrypted handshake message
server responds with Bad Record MAC alert

Maybe I forgot some synchronization, which causes these timeouts/unsupported Hearbeat (because I'm trying to communicate with server from another test)? I really don't see, what I missed in this.

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -945,3 +997,44 @@ def _changeWriteState(self):
+
+    def sendHeartbeatRequest(self, payload, padding_length):
+        """Sends heartbeat_request and saves payload from this message for
+           future checks of heartbeat responses.
```
that's not true, it's just synchronous version of `writeHeartbeat`

tomato42 commented on tlslite/tlsconnection.py:
```diff
@@ -1730,6 +1758,15 @@ def _handshakeServerAsyncHelper(self, verifierDB,
                 [ECPointFormat.uncompressed]))
 
+        # if client sent Heartbeat extension
+        if clientHello.getExtension(ExtensionType.heartbeat):
```
why the value of the client sent extension is ignored here?

mildass replied:
Value of the client sent extension is checked in '_serverGetClientHello' function (at line 2615) and after check all attributes (if we can send/receive etc.) are set. 
And here, in mentioned line, is just being checked, if client wants Heartbeat. If he wants it, then we send it too.

tomato42 commented:
> Maybe I forgot some synchronization, which causes these timeouts/unsupported Hearbeat (because I'm trying to communicate with server from another test)? I really don't see, what I missed in this.

the failure happens in `writeHeartbeat` and it happens because the values aren't set during handshake, you need to debug why they haven't been set; check if the extensions were sent and if they were sent with correct values, check if both peers acted on values in those extensions

tomato42 commented:
travis won't run until the merge conflicts are resolved, please rebase

tomato42 commented:
This pull request **introduces 1 alert** when merging 8c4ea95d6385948cc587e6f4de4b81d91c54fdc0 into ac9c99a59c3c51abbd6909e76415c6306c3e45d8 - [view on LGTM.com](https://lgtm.com/projects/g/tomato42/tlslite-ng/rev/pr-20d53d91e35f44126cddbdee5c1c73a8dd2c8e5f)

**new alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mildass commented:
Can you please rerun travis-ci? It failed because of some problem with `curl`. Thanks

tomato42 commented on tlslite/tlsrecordlayer.py:
```diff
@@ -163,6 +163,21 @@ def __init__(self, sock):
 
+        # Indicator for heartbeat extension mode, if we can receive
+        # heartbeat requests
+        self.heartbeatCanReceive = False
```
please rename those new variables to not use camelCase

tomato42 commented:
>  Coverage decreased (-1.3%) to 83.512% 

that's a large change, please inspect why that is, if necessary, add new tests to provide at least the same level of coverage as the rest of code has

tomato42 commented:
This pull request **introduces 1 alert** when merging 97e5be2706baef89dc31606dfec4c01c023c4246 into d85eb752aeb37bc5b02834570d3cb5bab4c8a9df - [view on LGTM.com](https://lgtm.com/projects/g/tomato42/tlslite-ng/rev/pr-425376daa7f711d3d75c85f1ced8b1350056f483)

**new alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mildass commented:
Sorry for missed `testConnServer` and `testConnClient`. I added it but after some rebase to master I had to delete it and then forget about it. It should be ok now.
I'd like to ask about camel case - where should be camelCase used? On every name (= attributes, local variables, class names etc.)? I renamed only attributes and local variables. Wasn't sure about class names and constants.

tomato42 commented:
> I'd like to ask about camel case - where should be camelCase used? On every name (= attributes, 
> local variables, class names etc.)? I renamed only attributes and local variables. Wasn't sure about 
> class names and constants.

camelCase should be used for class names, attributes and local variables should use `_` as separators and no upper case letters at all, constants should use `_` and only upper case letters

see the report from pylint for correct names

> Which tests 'counts' to tests coverage? tlstest.py? unit tests?

both, but unit tests are preferred as they are much faster to run

tomato42 commented:
@mildass what's the progress?

mildass commented:
@tomato42 Sorry, I forgot to push it (but I have done only renaming). 
Where I can find % tests coverage? (there was some check at PR, that showed me test coverage and I could see which parts aren't covered, but I don't see that now)

About renaming - you wrote, that class names should be camelCase and I get it as the first letter should be lowercase, but I was wrong and it should be uppercase, right?

About travis-ci jobs - why some jobs fail and some pass? I'm trying to find out why it is, but `make test-dev` is passing on my pc so I can't reproduce it.

tomato42 commented:
> About renaming - you wrote, that class names should be camelCase and I get it as the first letter should be lowercase, but I was wrong and it should be uppercase, right?

class names should start with capitalised letter

as I said "see the report from pylint for correct names"

> About travis-ci jobs - why some jobs fail and some pass? I'm trying to find out why it is, but make test-dev is passing on my pc so I can't reproduce it.

I see tests semi-consistently failing after "Test 54 - Heartbeat extension response callback" so looks like there is some synchronisation issue in it, likely in the server, as it's the side that doesn't abort with exception

tomato42 commented:
This pull request **introduces 1 alert** when merging f8afacb4c9bdc311139aaf91a6226b29b14c1537 into bad2b98b2c382674f71aff617a9274e2a095951f - [view on LGTM.com](https://lgtm.com/projects/g/tomato42/tlslite-ng/rev/pr-5389d41014545db01bc7c1c1ba6c3d7c557cc571)

**new alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

---

*Comment posted by [LGTM.com](https://lgtm.com)*

tomato42 commented:
This pull request **introduces 1 alert** when merging ebbfbf1fc62c1b0121b33393076fdde15a7e067f into bad2b98b2c382674f71aff617a9274e2a095951f - [view on LGTM.com](https://lgtm.com/projects/g/tomato42/tlslite-ng/rev/pr-634b3667780b5f8b1d0d79364b380ac540ec3b43)

**new alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mildass commented:
> as I said "see the report from pylint for correct names"

I totally missed that part of comment, sorry. 

I changed names and I hope that I didn't missed something (because pylint output shows a lot of invalid names which aren't related to this PR). Only `messageType` (`Heartbeat` class) and `extData` (`HeartbeatExtension` class) property names I left so.


> Coverage decreased (-0.09%) to 84.972% 

I added few tests. Is this fine now?

tomato42 commented on tlslite/constants.py:
```diff
@@ -124,7 +124,8 @@ class ContentType(TLSEnum):
     handshake = 22
     application_data = 23
-    all = (20, 21, 22, 23)
+    HEARTBEAT = 24  # RFC 6520
```
all the other identifiers are lower case because their names come from RFCs, so for consistency this one should also be lowercase

tomato42 commented on tlslite/constants.py:
```diff
@@ -145,6 +146,7 @@ class ExtensionType(TLSEnum):
     ec_point_formats = 11  # RFC 4492
     srp = 12  # RFC 5054
     signature_algorithms = 13  # RFC 5246
+    HEARTBEAT = 15  # RFC 6520
```
same here, for consistency, it should be lower case

tomato42 commented on tlslite/constants.py:
```diff
@@ -365,6 +367,20 @@ class CertificateStatusType(TLSEnum):
+    """Types of heartbeat messages from RFC 6520"""
+
+    HEARTBEAT_REQUEST = 1
+    HEARTBEAT_RESPONSE = 2
```
same here, for consistency and to reflect naming from RFCs, the names should be lower case

tomato42 commented on tlslite/messages.py:
```diff
@@ -2133,3 +2127,69 @@ def parse(self, p):
+    def messageType(self):
+        """Format heartbeat message to human readable representation."""
+        return none_as_unknown(HeartbeatMessageType.toRepr(
+                             self.message_type), self.message_type)
```
nit: wrong indentation

tomato42 commented on tlslite/messages.py:
```diff
@@ -2133,3 +2127,69 @@ def parse(self, p):
+        return w.bytes
+
+    @property
+    def messageType(self):
```
1. this is an internal method so it should be private
2. it's using camelCase name

mildass replied:
With private you mean just underscores in name? `__message_type`__message_type right?

tomato42 replied:
yes, with underscore at the begining, no, not `__message_type__` or `__message_type`, double underscore on both sides is for magic methods, double underscore at the beginning is for methods that should be hidden from child classes, so it's not necessary here

tomato42 commented on tests/tlstest.py:
```diff
@@ -1638,6 +1653,22 @@ def server_bind(self):
     test_no += 1
 
+    print("Test {0} - Heartbeat extension response callback".format(test_no))
+    synchro.send(b'R')
```
server should send the synchro when it is ready, as in, just before calling connect() and handshakeServer()

this means the os.urandom() definitely should be _before_ synchro.send(), setting up the callbacks probably should be also before it

in other test cases we can be a bit more forgiving as there only few values are set, which takes µs, not, potentially, seconds the os.urandom() can take

tomato42 commented on tests/tlstest.py:
```diff
@@ -1638,6 +1653,22 @@ def server_bind(self):
+    settings = HandshakeSettings()
+    heartbeat_payload = os.urandom(100)
+    def heartbeat_response_check(message):
+        assert heartbeat_payload == message.payload
```
if it receives a response, it should set a guard variable that is then checked after connection close to verify that the response was received

(but the check definitely has to be after testConnServer() as we need to do few reads and writes before we can be reasonably sure that the client had time to answer)

mildass replied:
How can I get variable out of callback scope? I'm trying something like this:
```
receive_payload = None
def heartbeat_response_check(message):
    received_payload = message.payload
```
But `received_payload` in and out are different. Is any simple way, how can I achieve that callback changes that variable?

tomato42 replied:
> How can I get variable out of callback scope?

either by using [`global`](https://docs.python.org/2.0/ref/global.html) or by changing a field in an object, not the object itself, compare:
```python
val = []
def fun():
   val[:] = ['some', 'values']
   print("ID of the object after assignment: {0}".format(id(val)))
print("ID of the object before assignment: {0}".format(id(val)))
print(val)
fun()
print(val)
```
with
```python
val = []
def fun():
   val = ['some', 'values']
   print("ID of the object after assignment: {0}".format(id(val)))
print("ID of the object before assignment: {0}".format(id(val)))
print(val)
fun()
print(val)
```

mildass replied:
Thanks, I didn't know that such a thing changes objects scope. 
Which of proposed solutions do you prefer here?

tomato42 commented:
This pull request **introduces 1 alert** when merging 96ff137b71325e2412ac9d437d85cf4826bd2da0 into bad2b98b2c382674f71aff617a9274e2a095951f - [view on LGTM.com](https://lgtm.com/projects/g/tomato42/tlslite-ng/rev/pr-bf13d77565184eddb651ad77a96b8347c63e4d75)

**new alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mildass commented:
@tomato42 After implementation required changes in `tlstest.py` I found, that server often doesn't have time to receive response on his heartbeat request in test.
When I add next `testConnServer` + `testConnClient` (so both are called 2 times in Heartbeat test), then it's all good, server has more time to get response and test is passing.
What do you recommend as solution? Dunno, if 2 times testConn is good, so I've been thinking about active waiting, which would just wait until global variable is set.

tomato42 commented:
> What do you recommend as solution? Dunno, if 2 times testConn is good, so I've been thinking about active waiting, which would just wait until global variable is set.

what if you send the requests, do the synchronisation read/write (`synchro.recv(1)`, `synchro.send(b'R')` – additional one, the one before `connect()` should remain), then `testConnServer()`, `testConnClient()`, and only then verify that the responses were received?

if that doesn't work, we can do the test twice, not such a big problem

> Edit: I have added Heartbeat to EncryptedExtensions for TLS1.3. Should I add it to existing commits or create new commit about it?

it can be separate

tomato42 commented:
This pull request **introduces 1 alert** when merging 9411cd3e714a2589053c90626e370a44550acecc into bad2b98b2c382674f71aff617a9274e2a095951f - [view on LGTM.com](https://lgtm.com/projects/g/tomato42/tlslite-ng/rev/pr-1fcc87d76d19f4dfec177c08caf3053478f9f292)

**new alerts:**

* 1 for \_\_eq\_\_ not overridden when adding attributes

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mildass commented:
Updated. It should work for TLS 1.3 now.
But there's redundant code in `_clientGetServerHello` and `_clientTLS13Handshake`. One works with ServerHello extension and another with EncryptedExtensions message. Don't know if it's ok create new method (probably in `TLSConnection` class) to get rid of this duplicity. Makes sense for me as programmer (no duplicate code) but doesn't make sense for me from the perspective of TLSConnection class structure (methods here are parts, which do whole sending/reading messages).

tomato42 commented:
yes, the whole `tlsconnection.py` is a train-wreck

for now let's leave it as is, once all the major features of TLS 1.3 are in, I'll take a look at removing the duplication and major refactoring

tomato42 commented:
looks good, thanks!

mildass commented:
Thanks for help, reviews and all answers to my questions!
